os/pm/pm_procfs.c

```diff
@@ -447,43 +460,117 @@ static size_t power_devices_read(FAR struct file *filep, FAR char *buffer, size_
+	2a. Else, we can consider to lock/unlock for all state other than PM_SLEEP, ie. fetch state: pm_querystate(PM_IDLE_DOMAIN)
+	2b. Need to verify if the transition is going to happen, but lock state appears suddenly, what will happen?
+	*/
+	pm_stay(PM_IDLE_DOMAIN, PM_NORMAL);
+	fvdbg("State locked!\n");
 
-	return buflen;
+	return OK;
 }
 
 /****************************************************************************
- * Name: power_unlock_write
+* 	Name: power_unlock_write
+*   
+* 	Description:
+*	Unlock a desired state by decreasing the counter of the state. A timer 
+*	interval can be provided to setup the timer interrupt. If there is some
+*	other interrupt appear before the interval ends, the system will also
+*	wakeup.
+*
+* 	Input Parameters:
+*   filep          - File path of the power domain
+*   buffer         - Desired state to be unlocked
+*   timer_interval - Timer interval for the timer interrupt
+*
+* 	Returned Value:
+*   0 (OK) means the state unlock was successful.
+*	1 (Failed) means the input state was incorrect.
  ****************************************************************************/
+#ifdef CONFIG_ARCH_CORTEXA32
+static size_t power_unlock_write(FAR struct file *filep, FAR const char *buffer, size_t timer_interval)
+#else
 static size_t power_unlock_write(FAR struct file *filep, FAR const char *buffer, size_t buflen)
+#endif
 {
-	int state = atoi(buffer);
-
-	if (state >= PM_NORMAL && state < PM_SLEEP) {
-		pm_relax(0, state);
-		fvdbg("power unlocked!\n");
-	} else {
-		fdbg("ERROR: Invalid state passed \n");
+	(void)buffer;
+
+	pm_relax(PM_IDLE_DOMAIN, PM_NORMAL);
+	fvdbg("State unlocked!\n");
+#ifdef CONFIG_ARCH_CORTEXA32
+	if (timer_interval > 0) {
+		g_timer_wakeup.use_timer = 1;
+		g_timer_wakeup.timer_interval = timer_interval;
 	}
-
+	else {
+		g_timer_wakeup.use_timer = 0;
+		g_timer_wakeup.timer_interval = 0;	
+	}
+	return OK;
+#else
 	return buflen;
+#endif
+}
+
+/****************************************************************************
+* 	Name: power_tunefreq_write
+*   
+* 	Description:
+*	Change the operating frequency of the AP core, for saving up power
+*	consumption under active mode.
+*
+* 	Input Parameters:
+*   filep            - File path of the power domain
+*   buffer(NOT USED) - Desired state to be unlocked
+*   div_lvl          - Level selection for operating frequency
+*
+* 	Returned Value:
+*   0 (OK) means the change of operating frequency was successful
+ ****************************************************************************/
+#ifdef CONFIG_ARCH_CORTEXA32
+static size_t power_tunefreq_write(FAR struct file *filep, FAR const char *buffer, size_t div_lvl)
+{
+	/* Buffer string need not to be considered, just div level is enough */
+	/*
+	   Only affect active mode power consumption
+	   0 -> 1.2GHz
+	   1 -> 600MHz
+	   2 -> 400Mhz
+	   3 -> 300Mhz
+	 */
+	(void)buffer;
+	SOCPS_Tune_APFreq(div_lvl);
```

Comment:
@amogh-samsung @edwakuwaku This is the chip-specific API. We don't call it directly from the common module. Could you check it?

Reply:
@amogh-samsung Even we add this like this now, we should clean up this. PM should not have any chip specific code. PM should have HAL or driver for this and chipset should enable the pm feature if it supports.

Reply:
Hi @sunghan-chang ,

Previously we have created another file named pm_ioctl.c, and using that as an interface to connect application and PM driver. But, I was told that that is not needed, which we should utilize the procfs layer for application control, which we then switch to verified that on test suite already.
As the requirement is keep on changing over time, I hope that Samsung side can have internal discussion regarding the application control design, and provide a unified and stable requirement.

Thank you.

Reply:
Hi,
I think the issue here is that though we need to use procfs only for syscalls, but we cannot invoke chipset specific APIs from here.
So we need to add an indirect call from here.

Reply:
https://github.com/amogh-samsung/TizenRT/commit/e2714de14628ddb41438d4dc6d720abde8433192

Please refer to this for a wrapped/mask call to chipset specific frequency tuning API